src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -0,0 +1,259 @@
+                                      const mfem::DeviceTensor<4, const double> J, int num_elements)
+{
+  using test_element          = finite_element<g, test>;
+  using trial_element         = finite_element<g, trial>;
+  using element_residual_type = typename trial_element::residual_type;
+  static constexpr auto rule  = GaussQuadratureRule<g, Q>();
+
+  const int grid_stride = blockDim.x * gridDim.x;
+#pragma unroll
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x; e < num_elements; e += grid_stride) {
+    // get the (change in) values for this particular element
+    tensor du_elem = detail::Load<trial_element>(du, e);
+
+    // this is where we will accumulate the (change in) element residual tensor
+    element_residual_type dr_elem{};
+
+    // for each quadrature point in the element
+    for (int q = 0; q < static_cast<int>(rule.size()); q++) {
+      gradient_quadrature<g, test, trial, Q, derivatives_type>(e, q, du_elem, dr_elem, derivatives_ptr, J,
+                                                               num_elements);
+    }
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(dr, dr_elem, e);
+  }
+}
+
+template <Geometry g, typename test, typename trial, int Q, typename derivatives_type, typename dsolution_type,
+          typename dresidual_type>
+__global__ void gradient_cuda_quadrature(const dsolution_type du, dresidual_type dr, derivatives_type* derivatives_ptr,
+                                         const mfem::DeviceTensor<4, const double> J, int num_elements)
+{
+  using test_element          = finite_element<g, test>;
+  using trial_element         = finite_element<g, trial>;
+  using element_residual_type = typename trial_element::residual_type;
+  static constexpr auto rule  = GaussQuadratureRule<g, Q>();
+
+  const int grid_stride           = blockDim.x * gridDim.x;
+  auto      thread_id             = blockIdx.x * blockDim.x + threadIdx.x;
+  auto      num_quadrature_points = num_elements * rule.size();
+#pragma unroll
+  for (int qe = thread_id; qe < num_quadrature_points; qe += grid_stride) {
+    int e = qe / rule.size();
+    int q = qe % rule.size();
+    // get the (change in) values for this particular element
+    tensor du_elem = detail::Load<trial_element>(du, e);
+
+    // this is where we will accumulate the (change in) element residual tensor
+    element_residual_type dr_elem{};
+
+    gradient_quadrature<g, test, trial, Q, derivatives_type>(e, q, du_elem, dr_elem, derivatives_ptr, J, num_elements);
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(dr, dr_elem, e);
+  }
+}
+
+template <Geometry g, typename test, typename trial, int Q, serac::detail::ThreadExecutionPolicy policy,
+          typename derivatives_type>
```

Comment:
These long templates need to be in a very specific order to ease use.  should `derivatives_types` go right after `Q` like the other functions?

Reply:
I agree that the execution policy should be last, or removed completely (in the short term)

Reply:
Sorry the naming of the policy type is confusing. I've changed it to be more reflective of what it actually is. It's a non-deduced value.

Reply:
Is there a reason to keep `derivatives_type` after the thread parallelization strategy policy? I agree it should be next to `Q`.

Reply:
Sorry it's a little unclear, but at the moment, `derivatives_type` is the first (and only) deduced type, while the parallelization strategy is explicitly supplied. So `Geometry` to the parallelization enum are all specified, but the `derivatives_type` is deduced.